core/samples/metrics/main.cpp

```diff
@@ -0,0 +1,86 @@
+/// [Include metrics]
+#include <cubos/core/metrics.hpp>
+using cubos::core::Metric;
+using cubos::core::Metrics;
+/// [Include metrics]
+#include <thread>
+
+#include <cubos/core/log.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+
+/// [Profiling metrics]
+void myFunction()
+{
+    CUBOS_PROFILE();
+    std::this_thread::sleep_for(std::chrono::seconds(2));
+    // this function should take about 2000ms...
+}
+/// [Profiling metrics]
+
+int main()
+{
+    /// [Register metrics]
+    CUBOS_METRIC("a", 1);
+    CUBOS_METRIC("b", 2);
+    CUBOS_METRIC("a", 3);
+    CUBOS_METRIC("c", 3);
+    /// [Register metrics]
+
+    /// [Search metrics]
+    std::size_t seenCount{0};
+    std::string name{};
+
+    CUBOS_ASSERT(Metrics::readName(name, seenCount));
+    CUBOS_ASSERT(seenCount == 1);
+    CUBOS_ASSERT(name == "a");
+
+    CUBOS_ASSERT(Metrics::readName(name, seenCount));
+    CUBOS_ASSERT(seenCount == 2);
+    CUBOS_ASSERT(name == "b");
+
+    CUBOS_ASSERT(Metrics::readName(name, seenCount));
+    CUBOS_ASSERT(seenCount == 3);
+    CUBOS_ASSERT(name == "c");
+
+    CUBOS_ASSERT(!Metrics::readName(name, seenCount)); // no more new metrics!
+    /// [Search metrics]
+
+    /// [Getting values]
+    double value;
+    std::size_t offset = 0;
+
+    CUBOS_ASSERT(Metrics::readValue("a", value, offset));
+    CUBOS_ASSERT(value == 1.0);
+    CUBOS_ASSERT(offset == 1);
+
+    CUBOS_ASSERT(Metrics::readValue("a", value, offset));
+    CUBOS_ASSERT(value == 3.0);
+    CUBOS_ASSERT(offset == 2);
+
+    offset = 0; // use a different offset or reset current one to 0
+    CUBOS_ASSERT(Metrics::readValue("b", value, offset));
+    CUBOS_ASSERT(value == 2.0);
+    CUBOS_ASSERT(offset == 1);
+
+    // etc..
+    /// [Getting values]
+
+    /// [Profiling metrics 2]
+    myFunction();
+
+    offset = 0;
+    CUBOS_ASSERT(Metrics::readValue("myFunction", value, offset));
```

Comment:
Won't this fail now, due to __FILE__ + __LINE__? Maybe you can get the name through the `readName` function.